### ISSUE TITLE
Implement an automated way of updating existing data

### ISSUE THREAD

**Issue by subotic** [dhlab-basel/Knora#792]
Following the discussion with @benjamingeer , we came to the conclusion that it would be a good idea to have an easy way of updating existing data, in the event of data-model changes. Currently, the system administrator would need to export the existing data and manually or via the `transform-data` application make the necessary changes. A very manual process, which leads to high reluctance to making necessary changes to the data-model (`knora-base` and co.), just so that the current user-base is not annoyed.

The idea is to have a version stamp in each data graph. When a new version of `webapi` is started, all data graphs would be scanned for this version stamp, and if the version found is old, then an automatic data conversion process would be launched. Beforehand, a backup of the old state would also be created.

**Comment by tobiasschweizer** [comment]
You will need to come up with a way to ensure that all the information is still there and the data did not get corrupted. 

I had to transform my data quite often and I always had to check that nothing got corrupted or lost my doing count queries in SPARQL and by actually looking at the output in SALSAH.

**Comment by subotic** [comment]
yes, I will think of something.

**Comment by subotic** [comment]
@benjamingeer I'm not sure that this is feasible if we only version the data graphs. The only safe way I see that could work (still probably not for all cases) is if we look at the whole repository (ontologies and data) as having a single version. Only then, when a user (system administrator) starts up a new version of `webapi` connected to a triplestore with an older version repository, we could consistently update the ontologies and the data.

Still, there are some cases where manual edits will probably still be necessary. For example, when there is a change in `knora-base` property. How are we going to migrate project specific sub-properties of those properties? The same thing with classes and sub-classes.

Then the question arises, is `knora-base` allowed to change, or are only new versions of `knora-base` allowed? So then we would have the version of the whole repository which would consist of specific versions of the single system ontologies. My head is going to explode 🤯

**Comment by benjamingeer** [comment]
> is if we look at the whole repository (ontologies and data) as having a single version

That sounds OK to me.

> For example, when there is a change in `knora-base` property. How are we going to migrate project specific sub-properties of those properties?

Suppose `knora-base:isPartOf` was renamed to `knora-base:isComponentOf`. Then you would just have to replace all triples like this:

```
example:examplePartOf rdfs:subPropertyOf knora-base:isPartOf .
```

with triples like this:

```
example:examplePartOf rdfs:subPropertyOf knora-base:isComponentOf .
```

But there's no need to handle this as a special case. Just delete all the triples that have the old property IRI as a subject, predicate, or object, and replace them with triples that have the new property IRI.

**Comment by benjamingeer** [comment]
> Then the question arises, is `knora-base` allowed to change, or are only new versions of `knora-base` allowed?

I think what we're talking about here is  having releases of `knora-base`. The repository would be tagged with the `knora-base` release number it uses. Then when `knora-base` gets a new release, the repository would have to be upgraded.

**Comment by loicjaouen** [comment]
There are a different ontologies, each data graph should point to a version of the ontologies it depends upon and we could have a dependency tree with update scheme.

The ontologies could be seen as APIs and released with a [sementic versionning](https://semver.org/) with X.Y.Z:
- Z: means no change on the data is required
  - for example, additions or internal change in the knora-base onto with border effects
  - responsibility: dasch
- Y: automatic changes possible
  - for example: like @benjamingeer 's example of renaming a property
  - responsibility: release manager, admin of the installation of knora
- X: editing is required, a decision is needed by the project owner
  - for example: the recent change of gui-order, in some project a single value was used on different resources using that property, it is moved to two values, by stretching this example a bit, one can imagine that the a choice might have to be done on the new value
  - responsible: the project's owner

**Comment by benjamingeer** [comment]
> There are different ontologies, each data graph should point to a version of the ontologies it depends upon

I think we can't do that, because there can only be one version of `knora-base` in the triplestore at a time. So you can't have different graphs with different versions. The whole triplestore has to be upgraded at the same time.

**Comment by loicjaouen** [comment]
Fully agree, still:
- _there can only be one version of knora-base in the triplestore at a time_, but there are also other ontologies (like `salsah-gui`)
- _The whole triplestore has to be upgraded at the same time_ but for the process to be automated, you have to know the steps by reading where you start from

**Comment by benjamingeer** [comment]
> but there are also other ontologies (like salsah-gui)

Yes, but they all have to work together. I think we should have releases of "the Knora ontologies" as a package.

> but for the process to be automated, you have to know the steps by reading where you start from

This is why Ivan is suggesting that we should tag the repository with the knora-ontologies version it uses.

> editing is required, a decision is needed by the project owner, for example: the recent change of gui-order

The automated update just copied the existing `guiOrder` on the property to every cardinality for that property. The result should be the same.

I hope will will never change built-in Knora ontologies in ways that would require *manual* changes in project-specific ontologies or data. So far we haven't had any such changes. Instead I think we could deprecate features we no longer want, while introducing new ones that are backwards-compatible.

Let's try to keep this simple:

* The package of built-in Knora ontologies has releases and version numbers.
* The repository as a whole contains a triple specifying its knora-ontologies version.
* There is an automated upgrade process for moving between each version and the subsequent one.
* The automated upgrade is done when the server starts:
    * It gives you the latest version of `knora-ontologies`.
    * It upgrades your data to be compatible with that version.
* If you skip versions, you get all the data upgrade steps between your version and the latest one.

One thing to keep in mind: In GraphDB, the upgrade will have to be done in a single transaction, otherwise consistency checks will fail. This means that the built-in Knora ontologies will also have to be updated in the same transaction (replacing the contents of their graphs with the new built-in ontologies). This should be easier now that RDF4J executes a sequence of update operations in a single transaction (https://github.com/eclipse/rdf4j/pull/856).

Sound OK?

**Comment by subotic** [comment]
Yes, sounds good.

**Comment by subotic** [comment]
wip in https://github.com/dhlab-basel/Knora/pull/788

**Comment by subotic** [comment]
As the first case, I'm using the changes to the data necessary for the separation of `knora-base` and `knora-admin`. The new Knora ontology package will be known as version `1`. This means that the current Knora ontologies in `develop` are implicitly version `0`. This also means that Lausanne would first need to manually upgrade to version `0`, before an automatic upgrade to `1` can happen.

**Comment by loicjaouen** [comment]
@subotic did we hear a ring? Sorry, just to be sure

> This means that the current Knora ontologies in `develop` are implicitly version `0`

do you assume that `develop` is knora [v1.2.0](https://github.com/dhlab-basel/Knora/releases/tag/v1.2.0)?

> This also means that Lausanne would first need to manually upgrade to version `0`

_manually_ means outside of the `webapi` auto upgrade mechanism?

If it is so, then:
- knora [v1.2.0](https://github.com/dhlab-basel/Knora/releases/tag/v1.2.0) depends on a set of _knora base ontologies_ that are tagged `v0`
- you develop an ecosystem of _knora base ontologies_ `v1` for a coming knora release (probably `v1.3.0`), which would have an automated way of upgrading its triplestore from `v0` to `v1`
- so we (Lausanne) should take the required steps to upgrade to [v1.2.0](https://github.com/dhlab-basel/Knora/releases/tag/v1.2.0) first

**Comment by benjamingeer** [comment]
Aren't Lausanne already using the Knora ontologies that are in `develop`?

**Comment by loicjaouen** [comment]
We are in `v1.1.0` right now.

**Comment by subotic** [comment]
> do you assume that develop is knora v1.2.0?

no, everything before this PR is `0`, meaning currently from commit https://github.com/dhlab-basel/Knora/commit/cea8e000d01c3d02f4b752bf0bd7731e494e2302.

> manually means outside of the webapi auto upgrade mechanism?

yes

I plan to merge this in April, so if no commits are added after tomorrow that need upgrading, then `v1.3.0` will be `0`. In the case that there are any commits after tomorrows release that need data upgrading, then I will first make a release and then merge the PR into `develop`. Then it should be clear what `0` and what `1` is going to be.
